compiler/rustc_typeck/src/check/place_op.rs

```diff
@@ -82,6 +86,53 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             expr, base_expr, adjusted_ty, index_ty
         );
 
+        let negative_index = || {
+            let ty = self.resolve_vars_if_possible(adjusted_ty);
+            let mut err = self.tcx.sess.struct_span_err(
+                idx_expr.span,
+                &format!("negative integers cannot be used to index on a `{}`", ty),
+            );
+            err.span_label(
+                idx_expr.span,
+                &format!("cannot use a negative integer for indexing on `{}`", ty),
+            );
+            if let (hir::ExprKind::Path(..), Ok(snippet)) =
+                (&base_expr.kind, self.tcx.sess.source_map().span_to_snippet(base_expr.span))
+            {
+                // `foo[-1]` to `foo[foo.len() - 1]`
+                err.span_suggestion_verbose(
+                    idx_expr.span.shrink_to_lo(),
+                    &format!(
+                        "if you wanted to access an element starting from the end of the `{}`, you \
```

Comment:
```suggestion
                        "to access an element starting from the end of the `{}`, compute the index"
```
nit: this wording is more concise